lib/iterator.js

```diff
@@ -183,6 +183,30 @@ class Iterator {
     return new SkipWhileIterator(this, predicate, thisArg);
   }
 
+  // Consumes an iterator, partitioning it into Arrays.
+  // Signature: predicate[, thisArg]
+  //   predicate <Function> function returns a value to partition this iterator
+  //     value <any> current iterator element.
+  //     Returns: <boolean> | <number> key denoting resulting partition this
+  //         value will be assigned to. Number denotes index in the resulting
+  //         array. Boolean will be mapped `false` -> `0`, `true` -> `1`.
+  //   thisArg <any> value to be used as `this` when calling `predicate`
+  //   Returns: <Array> array of partitions (arrays), will always have at
+  //       least 2 arrays in it.
+  partition(predicate, thisArg) {
+    const result = [[], []];
+    for (const value of this) {
+      const res = predicate.call(thisArg, value);
+      let index;
+      if (typeof res !== 'number') index = res ? 1 : 0;
+      else index = res;
+      const arr = result[index];
+      if (!arr) result[index] = [value];
+      else arr.push(value);
```

Comment:
To improve performance:
```suggestion
      if (result.length <= index) {
        const oldLength = result.length;
        result.length = index + 1;
        for (let i = oldLength; i < result.length; i++) result[i] = [];
      }
      result[index].push(value);
```